pkg/controller/job/success_policy.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package job
+
+import (
+	"fmt"
+
+	batch "k8s.io/api/batch/v1"
+	"k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/features"
+)
+
+func canApplySuccessPolicy(jobCtx *syncJobCtx) bool {
+	return isIndexedJob(jobCtx.job) && feature.DefaultFeatureGate.Enabled(features.JobSuccessPolicy) &&
+		jobCtx.failureCondition == nil && jobCtx.successCondition == nil && jobCtx.job.Spec.SuccessPolicy != nil
+}
+
+func metSuccessPolicy(logger klog.Logger, successPolicy *batch.SuccessPolicy, completions int32, succeededIndexes orderedIntervals) (string, bool) {
+	if successPolicy == nil || len(succeededIndexes) == 0 {
+		return "", false
+	}
+
+	criteriaMetMsg := "Met criteria at index"
+	for index, criteria := range successPolicy.Criteria {
+		if criteria.SucceededIndexes != nil {
+			requiredIndexes := parseIndexesFromString(logger, *criteria.SucceededIndexes, int(completions))
+			// Failed to parse succeededIndexes of the criteria due to some errors like invalid format.
+			if len(requiredIndexes) == 0 {
+				continue
+			}
+			if metSucceededIndexesCriteria(requiredIndexes, succeededIndexes, criteria.SucceededCount) {
+				return fmt.Sprintf("%s %d", criteriaMetMsg, index), true
+			}
+		} else if criteria.SucceededCount != nil && succeededIndexes.total() >= int(*criteria.SucceededCount) {
+			return fmt.Sprintf("%s %d", criteriaMetMsg, index), true
+		}
+	}
+	return "", false
+}
+
+func metSucceededIndexesCriteria(requiredIndexes, succeededIndexes orderedIntervals, succeededCount *int32) bool {
+	var contains, succeededPointer int
+	for _, criteria := range requiredIndexes {
+		for succeededPointer < len(succeededIndexes) {
+			if overlap := min(criteria.Last, succeededIndexes[succeededPointer].Last) - max(criteria.First, succeededIndexes[succeededPointer].First) + 1; overlap > 0 {
+				contains += overlap
+			}
+			if succeededIndexes[succeededPointer].Last < criteria.Last {
+				succeededPointer++
+			} else if succeededIndexes[succeededPointer].Last == criteria.Last {
+				succeededPointer++
+				break
```

Comment:
TBH, I find this `break`s hard to read, maybe instead we could have two pointers, one for criteria, one for succeeded. WDYT:
```
func metSucceededIndexesCriteria(requiredIndexes, succeededIndexes orderedIntervals, succeededCount *int32) bool {
	var contains, succeededPointer, criteriaPointer int
	for criteriaPointer < len(requiredIndexes) && succeededPointer < len(succeededIndexes) {
		criteriaInterval, successInterval := requiredIndexes[criteriaPointer], succeededIndexes[succeededPointer]
		if overlap := min(criteriaInterval.Last, successInterval.Last) - max(criteriaInterval.First, successInterval.First) + 1; overlap > 0 {
			contains += overlap
		}
		if successInterval.Last <= criteriaInterval.Last {
			succeededPointer++
		}
		if successInterval.Last >= criteriaInterval.Last {
			criteriaPointer++
		}
	}
	return contains == requiredIndexes.total() || (succeededCount != nil && contains >= int(*succeededCount))
}
```

Reply:
I'm ok with removing these `break`s. Let me check and verify if these new logic would work well.
Thanks.

Reply:
I think this should also work, and maybe is cleaner?
```
if successInterval.Last < criteriaInterval.Last {
  // the current succeeded interval is behind, so we can move to the next
  succeededPointer++
} else if  successInterval.Last > criteriaInterval.Last {
  // the current criteria interval is behind, so we can move to the next
  criteriaPointer++
} else {
  // both intervals end at the same position, we can move to the next succeeded, and next criteria
  succeededPointer++
  criteriaPointer++
}
``

Reply:
@mimowo In the case of your latest suggestions, how to calculate `contains`?

Reply:
it is computed above as before, I only proposed another version of moving the pointers

Reply:
Sgtm

Reply:
Done.